File: ios/Classes/TobiasPlugin.m
#import "TobiasPlugin.h"
#import <AlipaySDK/AlipaySDK.h>

__weak TobiasPlugin* __tobiasPlugin;

@interface TobiasPlugin()

@property (readwrite,copy,nonatomic) FlutterResult callback;

@end



@implementation TobiasPlugin

-(id)init{
    if(self = [super init]){
        
        __tobiasPlugin  = self;
        
    }
    return self;
}

-(void)dealloc{
    
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"com.jarvanmo/tobias"
            binaryMessenger:[registrar messenger]];
  TobiasPlugin* instance = [[TobiasPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
  [registrar addApplicationDelegate:instance];
}




- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"pay" isEqualToString:call.method]) {
      [self pay:call result:result];
  } else if([@"version" isEqualToString:call.method]){
      [self getVersion:call result:result];
  } else if([@"auth" isEqualToString:call.method]){
      [self _auth:call result:result];
  } else if([@"isAliPayInstalled" isEqualToString:call.method]){
      [self  _isAliPayInstalled:call result:result];
  }else{
      result(FlutterMethodNotImplemented);
  }
}

- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceApplication:(NSString *)sourceApplication annotation:(id)annotation {

    return [self handleOpenURL:url];
}

// NOTE: 9.0以后使用新API接口
- (BOOL)application:(UIApplication *)app openURL:(NSURL *)url options:(NSDictionary<NSString *, id> *)options {
    return [self handleOpenURL:url];
}


+(BOOL)handleOpenURL:(NSURL*)url{
  
    if(!__tobiasPlugin)return NO;
    return [__tobiasPlugin handleOpenURL:url];
    
}


-(BOOL)handleOpenURL:(NSURL*)url{
    
    if ([url.host isEqualToString:@"safepay"]) {

        __weak TobiasPlugin* __self = self;

        
        [[AlipaySDK defaultService] processOrderWithPaymentResult:url standbyCallback:^(NSDictionary *resultDic) {
             [__self onPayResultReceived:resultDic];
        }];

        [[AlipaySDK defaultService] processAuth_V2Result:url standbyCallback:^(NSDictionary *resultDic) {
             [__self onAuthResultReceived:resultDic];
         }];

        return YES;
    }
    return NO;
}

-(void)onPayResultReceived:(NSDictionary*)resultDic{

    if(self.callback!=nil){
        NSMutableDictionary *mutableDictionary = [NSMutableDictionary dictionaryWithDictionary:resultDic];
        [mutableDictionary setValue:@"iOS" forKey:@"platform"];
        self.callback(mutableDictionary);
        self.callback = nil;
    }
    
}

-(void)onAuthResultReceived:(NSDictionary*)resultDic{

    if(self.callback!=nil){
        NSMutableDictionary *mutableDictionary = [NSMutableDictionary dictionaryWithDictionary:resultDic];
        [mutableDictionary setValue:@"iOS" forKey:@"platform"];
        self.callback(mutableDictionary);
        self.callback = nil;
    }

}

-(void) pay:(FlutterMethodCall*)call result:(FlutterResult)result{

    NSString* urlScheme = [self fetchUrlScheme];
    if(!urlScheme){
        result([FlutterError errorWithCode:@"AliPay UrlScheme Not Found" message:@"Config AliPay First" details:nil]);
        return;
    }

 
    [self _pay:call result:result urlScheme:urlScheme];

}

-(void) _pay:(FlutterMethodCall*)call result:(FlutterResult)result urlScheme:(NSString *)urlScheme{
    self.callback = result;
    
    __weak TobiasPlugin* __self = self;

    [[AlipaySDK defaultService] payOrder:call.arguments[@"order"] fromScheme:urlScheme callback:^(NSDictionary *resultDic) {
        [__self onPayResultReceived:resultDic];
    }];

}

-(void) _auth:(FlutterMethodCall*)call result:(FlutterResult)result {
    
    NSString* urlScheme = [self fetchUrlScheme];
    if(!urlScheme){
        result([FlutterError errorWithCode:@"AliPay UrlScheme Not Found" message:@"Config AliPay First" details:nil]);
        return;
    }
    self.callback = result;
    
    __weak TobiasPlugin* __self = self;
    
  [[AlipaySDK defaultService] auth_V2WithInfo:call.arguments
                                         fromScheme:urlScheme
                                           callback:^(NSDictionary *resultDic) {
                                               [__self onAuthResultReceived:resultDic];
                                           }];
}

-(void) getVersion:(FlutterMethodCall*)call result:(FlutterResult)result{

    NSString *version = [AlipaySDK defaultService].currentVersion;
    if(version == nil){
        version = @"";
    }
    result(version);
}

-(NSString*)fetchUrlScheme{
    NSDictionary *infoDic = [[NSBundle mainBundle] infoDictionary];
    NSArray* types = infoDic[@"CFBundleURLTypes"];
    for(NSDictionary* dic in types){
        if([@"alipay" isEqualToString:dic[@"CFBundleURLName"]]){
            return dic[@"CFBundleURLSchemes"][0];
        }
    }
    return nil;
}

-(void) _isAliPayInstalled:(FlutterMethodCall*)call result:(FlutterResult)result {
   BOOL isAliPayInstalled = [[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:@"alipays://"]]||[[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:@"alipay://"]];
   result(@(isAliPayInstalled));
}

@end

File: ios/tobias/Sources/tobias/TobiasPlugin.swift
import Flutter
import UIKit
import AlipaySDK

@objc public class TobiasPlugin: NSObject, FlutterPlugin, FlutterApplicationLifeCycleDelegate {
    private static weak var instance: TobiasPlugin?
    private var callback: FlutterResult?
    
    public override init() {
        super.init()
        TobiasPlugin.instance = self
    }
    

    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.jarvanmo/tobias",
            binaryMessenger: registrar.messenger()
        )
        let instance = TobiasPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "pay":
            pay(call: call, result: result)
        case "registerApp":
            registerApp(call: call, result: result)
        case "version":
            getVersion(call: call, result: result)
        case "auth":
            auth(call: call, result: result)
        case "isAliPayInstalled":
            isAliPayInstalled(call: call, result: result)
        case "isAliPayHKInstalled":
            isAliPayHKInstalled(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
        
    @objc public func application(
        _ application: UIApplication,
        open url: URL,
        options: [UIApplication.OpenURLOptionsKey: Any] = [:]
    ) -> Bool {
        return handleOpenURL(url: url)
    }
    
    @objc public func application(
        _ application: UIApplication,
        open url: URL,
        sourceApplication: String?,
        annotation: Any
    ) -> Bool {
        return handleOpenURL(url: url)
    }
    
    public func application(
        _ application: UIApplication,
        continue userActivity: NSUserActivity,
        restorationHandler: @escaping ([Any]) -> Void
    ) -> Bool {
        if userActivity.activityType == NSUserActivityTypeBrowsingWeb {
            AlipaySDK.defaultService().handleOpenUniversalLink(userActivity) { (resultDic) in
                // Handle universal link callback if needed
            }
        }
        return false
    }
    
    // MARK: - Static URL Handler
    
    @objc public static func handleOpenURL(url: URL) -> Bool {
        guard let instance = instance else {
            return false
        }
        return instance.handleOpenURL(url: url)
    }
    
    // MARK: - Private Methods
    
    private func handleOpenURL(url: URL) -> Bool {
        guard url.host == "safepay" else {
            return false
        }
        
        AlipaySDK.defaultService().processOrder(withPaymentResult: url, standbyCallback: { [weak self] (resultDic) in
            self?.onPayResultReceived(resultDic: resultDic)
        })
        
        AlipaySDK.defaultService().processAuth_V2Result(url, standbyCallback: { [weak self] (resultDic) in
            self?.onAuthResultReceived(resultDic: resultDic)
        })
        
        return true
    }
    
    private func onPayResultReceived(resultDic: [AnyHashable: Any]?) {
        guard let callback = callback else {
            return
        }
        
        guard let resultDic = resultDic else {
            self.callback = nil
            return
        }
        
        var mutableDictionary: [String: Any] = [:]
        for (key, value) in resultDic {
            if let stringKey = key as? String {
                mutableDictionary[stringKey] = value
            } else {
                mutableDictionary[String(describing: key)] = value
            }
        }
        mutableDictionary["platform"] = "iOS"
        callback(mutableDictionary)
        self.callback = nil
    }
    
    private func onAuthResultReceived(resultDic: [AnyHashable: Any]?) {
        guard let callback = callback else {
            return
        }
        
        guard let resultDic = resultDic else {
            self.callback = nil
            return
        }
        
        var mutableDictionary: [String: Any] = [:]
        for (key, value) in resultDic {
            if let stringKey = key as? String {
                mutableDictionary[stringKey] = value
            } else {
                mutableDictionary[String(describing: key)] = value
            }
        }
        mutableDictionary["platform"] = "iOS"
        callback(mutableDictionary)
        self.callback = nil
    }
    
    private func pay(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let urlScheme = fetchUrlScheme() else {
            result(FlutterError(
                code: "AliPay UrlScheme Not Found",
                message: "Config AliPay First",
                details: nil
            ))
            return
        }
        
        pay(call: call, result: result, urlScheme: urlScheme)
    }
    
    private func pay(call: FlutterMethodCall, result: @escaping FlutterResult, urlScheme: String) {
        self.callback = result
        
        guard let arguments = call.arguments as? [String: Any],
              let order = arguments["order"] as? String else {
            result(FlutterError(
                code: "InvalidArguments",
                message: "Order string is required",
                details: nil
            ))
            self.callback = nil
            return
        }
        
        let universalLink = arguments["universalLink"] as? String
        
        AlipaySDK.defaultService().payOrder(
            order,
            fromScheme: urlScheme,
            fromUniversalLink: universalLink
        ) { [weak self] (resultDic) in
            self?.onPayResultReceived(resultDic: resultDic)
        }
    }
    
    private func auth(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let urlScheme = fetchUrlScheme() else {
            result(FlutterError(
                code: "AliPay UrlScheme Not Found",
                message: "Config AliPay First",
                details: nil
            ))
            return
        }
        
        self.callback = result
        
        
        
        guard let arguments = call.arguments as? String else {
            result(FlutterError(
                code: "InvalidArguments",
                message: "Auth info is required",
                details: nil
            ))
            self.callback = nil
            return
        }
        AlipaySDK.defaultService().auth_V2(
            withInfo: arguments,
            fromScheme: urlScheme
        ) { [weak self] (resultDic) in
            self?.onAuthResultReceived(resultDic: resultDic)
        }
    }
    
    private func getVersion(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let version = AlipaySDK.defaultService().currentVersion
        result(version)
    }
    
    private func registerApp(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let appId = arguments["appId"] as? String else {
            result(FlutterError(
                code: "InvalidArguments",
                message: "App ID is required",
                details: nil
            ))
            return
        }
        
        let universalLink = arguments["universalLink"] as? String
        AlipaySDK.defaultService().registerApp(appId, universalLink: universalLink)
        result(nil)
    }
    
    private func fetchUrlScheme() -> String? {
        guard let infoDic = Bundle.main.infoDictionary,
              let types = infoDic["CFBundleURLTypes"] as? [[String: Any]] else {
            return nil
        }
        
        for dic in types {
            if let urlName = dic["CFBundleURLName"] as? String,
               urlName == "alipay",
               let schemes = dic["CFBundleURLSchemes"] as? [String],
               let firstScheme = schemes.first {
                return firstScheme
            }
        }
        
        return nil
    }
    
    private func isAliPayInstalled(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let alipaysURL = URL(string: "alipays://"),
              let alipayURL = URL(string: "alipay://") else {
            result(false)
            return
        }
        
        let canOpenAlipays = UIApplication.shared.canOpenURL(alipaysURL)
        let canOpenAlipay = UIApplication.shared.canOpenURL(alipayURL)
        result(canOpenAlipays || canOpenAlipay)
    }
    
    private func isAliPayHKInstalled(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let alipayhkURL = URL(string: "alipayhk://") else {
            result(false)
            return
        }
        
        let canOpen = UIApplication.shared.canOpenURL(alipayhkURL)
        result(canOpen)
    }
}

